try to refine type

bernt-matthias opened the pull request:
by checking for subclasses

bernt-matthias committed:
try to refine type

by checking for subclasses

multimeric commented:
This seems like a reasonable approach. I do have some feedback, however.

multimeric commented on acclimatise/model.py:
```diff
@@ -293,15 +293,21 @@ def _name_from_arg(self) -> typing.Iterable[str]:
-
-        return infer_type(self.description) or cli_types.CliString()
+        tpe = None
+        tpe_cand = [
```
I prefer descriptive variable names where possible. `typ` and `type_candidates` please.

bernt-matthias replied:
> I would have though your example in #37 would predict None, then CliFile() then CliFile() respectively, and then return CliFile(). We only use subclasses for CliFileSystemType, and only recently, but the inference will never return CliFileSystemType, so it seems irrelevant.

True. I already forgot that we implemented the distinction between in/output files/dirs by properties and not subclasses.Then I will check for `isinstance`.

multimeric replied:
Ah that makes more sense

multimeric commented:
Out of interest, what is an example of this PR improving behaviour? I would have though your example in #37 would predict `None`, then `CliFile()` then `CliFile()` respectively, and then return `CliFile()`. We only use subclasses for `CliFileSystemType`, and only recently, but the inference will never return `CliFileSystemType`, so it seems irrelevant.

If I ever rewrite my type system (and the `representable` property), I might be able to make this work better.

multimeric reviewed (changes requested):
Could you please add a test case also please. You should be able to use the exact example from the linked issue.

bernt-matthias commented:
> Could you please add a test case also please. You should be able to use the exact example from the linked issue.

Can try. Do you have an example how this is done? Or where?

bernt-matthias committed:
use isinstance instead of issubclass

since in/out file/dir is represented by properties of the same class

- also fixes linter errors

multimeric commented:
Ah well eventually I'll make a `test_data_model.py` or something, but in the meantime I think you could chuck this into `test_type_inference.py`. You can construct a `Flag()` object with the fields that you mentioned in #37, and then assert that the value of `Flag.get_type()` returns the right value.

multimeric commented:
Also look into getting `pre-commit` set-up so that the builds pass, and the git diffs are easier for me to read. You should just have to run:
```bash
pip install pre-commit
pre-commit install
```
